### ISSUE TITLE
rusqlite: without modern_sqlite, some tests fail

### ISSUE THREAD

**Issue by dkg** [rusqlite/rusqlite#1216]
I'm trying to automate the build process for a non-bundled version of the `rusqlite` crate (and its `libsqlite3-sys` dependency) for debian.  Debian ships an independent copy of `libsqlite`, so bundling isn't necessary.  Part of the debian packaging process encourages me to run tests with individual features selectively enabled or disabled.

This shows that some of the tests seem to implicitly assume that they are being run with the `modern_sqlite` despite not being marked as such.

For example, when i run the tests with just `--features vtab`, i see the following error:

```
error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/vtablog.rs:118:14
    |
118 |         info.set_estimated_rows(500);
    |              ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`
```

According to the comments in `src/vtab/mod.rs`, `set_estimated_rows` was introduced in 3.8.2, which was from December 2013, 8.5 years ago.

Another similar mechanism that is hidden behind `modern_sqlite` is `VTabConfig`, introduced in sqlite 3.7.7, June 2011, over a decade ago.

I see two distinct possible ways of fixing this kind of problem:

- annotate each test that uses a part of the crate that depends on the `modern_sqlite` feature explicitly as such, so that the tests don't fail.
- bring some of the older features (like `VTabConfig` and `set_estimated_rows`) out of the `modern_sqlite` feature and into the default, featureless crate.  It seems very unlikely that anyone will build rusqlite in 2022 and beyond against a really ancient version of libsqlite3.

Here is a full listing of all the errors that i've seen that are tickled by minimal featuresets without the `modern_sqlite` feature:

```
error[E0432]: unresolved import `crate::vtab::VTabConfig`
  --> src/vtab/csvtab.rs:34:5
   |
34 |     VTabConfig, VTabConnection, VTabCursor, VTabKind, Values,
   |     ^^^^^^^^^^ no `VTabConfig` in `vtab`

error[E0432]: unresolved import `crate::vtab::VTabConfig`
  --> src/vtab/series.rs:13:73
   |
13 |     eponymous_only_module, Context, IndexConstraintOp, IndexInfo, VTab, VTabConfig, VTabConnection,
   |                                                                         ^^^^^^^^^^ no `VTabConfig` in `vtab`

    --> src/context.rs:31:25
     |
31   |             return ffi::sqlite3_result_pointer(
     |                         ^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `sqlite3_result_int`
     |
    ::: /home/dkg/tmp/tmp.XdFIONcPD6/target/x86_64-unknown-linux-gnu/debug/build/libsqlite3-sys-8ccc2a5a3d8e41de/out/bindgen.rs:2407:5
     |
2407 |     pub fn sqlite3_result_int(arg1: *mut sqlite3_context, arg2: ::std::os::raw::c_int);
     |     ----------------------------------------------------------------------------------- similarly named function `sqlite3_result_int` defined here

error[E0425]: cannot find function `sqlite3_create_window_function` in crate `ffi`
    --> src/functions.rs:588:18
     |
588  |               ffi::sqlite3_create_window_function(
     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `sqlite3_create_function`
     |
    ::: /home/dkg/tmp/tmp.XdFIONcPD6/target/x86_64-unknown-linux-gnu/debug/build/libsqlite3-sys-8ccc2a5a3d8e41de/out/bindgen.rs:2188:5
     |
2188 | /     pub fn sqlite3_create_function(
2189 | |         db: *mut sqlite3,
2190 | |         zFunctionName: *const ::std::os::raw::c_char,
2191 | |         nArg: ::std::os::raw::c_int,
...    |
2208 | |         xFinal: ::std::option::Option<unsafe extern "C" fn(arg1: *mut sqlite3_context)>,
2209 | |     ) -> ::std::os::raw::c_int;
     | |_______________________________- similarly named function `sqlite3_create_function` defined here

error[E0425]: cannot find function `sqlite3_bind_pointer` in crate `ffi`
    --> src/statement.rs:743:26
     |
743  |                       ffi::sqlite3_bind_pointer(
     |                            ^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `sqlite3_bind_int`
     |
    ::: /home/dkg/tmp/tmp.XdFIONcPD6/target/x86_64-unknown-linux-gnu/debug/build/libsqlite3-sys-8ccc2a5a3d8e41de/out/bindgen.rs:1969:5
     |
1969 | /     pub fn sqlite3_bind_int(
1970 | |         arg1: *mut sqlite3_stmt,
1971 | |         arg2: ::std::os::raw::c_int,
1972 | |         arg3: ::std::os::raw::c_int,
1973 | |     ) -> ::std::os::raw::c_int;
     | |_______________________________- similarly named function `sqlite3_bind_int` defined here

error[E0425]: cannot find function `sqlite3_value_pointer` in crate `ffi`
    --> src/vtab/mod.rs:764:33
     |
764  |         let ptr = unsafe { ffi::sqlite3_value_pointer(arg, array::ARRAY_TYPE) };
     |                                 ^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `sqlite3_value_int`
     |
    ::: /home/dkg/tmp/tmp.XdFIONcPD6/target/x86_64-unknown-linux-gnu/debug/build/libsqlite3-sys-8ccc2a5a3d8e41de/out/bindgen.rs:2304:5
     |
2304 |     pub fn sqlite3_value_int(arg1: *mut sqlite3_value) -> ::std::os::raw::c_int;
     |     ---------------------------------------------------------------------------- similarly named function `sqlite3_value_int` defined here

error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/array.rs:110:18
    |
110 |             info.set_estimated_rows(100);
    |                  ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`

error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/array.rs:114:18
    |
114 |             info.set_estimated_rows(2_147_483_647);
    |                  ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`

error[E0599]: no method named `config` found for mutable reference `&mut VTabConnection` in the current scope
   --> src/vtab/csvtab.rs:239:12
    |
239 |         db.config(VTabConfig::DirectOnly)?;
    |            ^^^^^^ method not found in `&mut VTabConnection`

error[E0599]: no method named `config` found for mutable reference `&mut VTabConnection` in the current scope
  --> src/vtab/series.rs:67:12
   |
67 |         db.config(VTabConfig::Innocuous)?;
   |            ^^^^^^ method not found in `&mut VTabConnection`

error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/series.rs:125:18
    |
125 |             info.set_estimated_rows(1000);
    |                  ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`

error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/series.rs:150:18
    |
150 |             info.set_estimated_rows(2_147_483_647);
    |                  ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`
error[E0599]: no method named `set_estimated_rows` found for mutable reference `&mut IndexInfo` in the current scope
   --> src/vtab/vtablog.rs:118:14
    |
118 |         info.set_estimated_rows(500);
    |              ^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `set_estimated_cost`
```

I could offer a patch to move at least `set_estimated_rows` and `VTabConfig` outside of `modern_sqlite`, and annotate the other tests appropriately (with `#[cfg(feature = "modern_sqlite")]`) as needed, if that seems like a reasonable approach.

**Comment by gwenn** [comment]
See #706 for the main reason why we need `modern_sqlite`.
A partial fix is available here: #1217.

**Comment by gwenn** [comment]
See #1221 for a fix.
